pilot/pkg/xds/eds.go

```diff
@@ -293,88 +294,126 @@ func connectionID(node string) string {
 	return node + "-" + strconv.FormatInt(id, 10)
 }
 
+// SubsetToLabels returns the labels associated with a subset of a given service.
+func subsetToLabels(dr *model.Config, subsetName string) labels.Collection {
+	// empty subset
+	if subsetName == "" {
+		return nil
+	}
+
+	if dr == nil {
+		return nil
+	}
+
+	rule := dr.Spec.(*networkingapi.DestinationRule)
+	for _, subset := range rule.Subsets {
+		if subset.Name == subsetName {
+			if len(subset.Labels) == 0 {
+				return nil
+			}
+			return []labels.Instance{subset.Labels}
+		}
+	}
+
+	return nil
+}
+
 // loadAssignmentsForClusterIsolated return the endpoints for a proxy in an isolated namespace
 // Initial implementation is computing the endpoints on the flight - caching will be added as needed, based on
 // perf tests. The logic to compute is based on the current UpdateClusterInc
-func (s *DiscoveryServer) loadAssignmentsForClusterIsolated(proxy *model.Proxy, push *model.PushContext,
-	clusterName string) *endpoint.ClusterLoadAssignment {
-	_, subsetName, hostname, port := model.ParseSubsetKey(clusterName)
-
-	// TODO: BUG. this code is incorrect if 1.1 isolation is used. With destination rule scoping
-	// (public/private) as well as sidecar scopes allowing import of
-	// specific destination rules, the destination rule for a given
-	// namespace should be determined based on the sidecar scope or the
-	// proxy's config namespace. As such, this code searches through all
-	// destination rules, public and private and returns a completely
-	// arbitrary destination rule's subset labels!
-	subsetLabels := push.SubsetToLabels(proxy, subsetName, hostname)
+func (s *DiscoveryServer) loadAssignmentsForClusterIsolated(b EndpointBuilder) *endpoint.ClusterLoadAssignment {
```

Comment:
while you are at it, should we rename this method `loadAssignmentsForClusterIsolated`? I was not clear what `Isolated` means here. If it has some meaning, please ignore this comment